ui/src/components/layout/Revisions.vue

```diff
@@ -0,0 +1,316 @@
+            emit("restore", revisionSource);
+            return Promise.resolve();
+        });
+    }
+
+    function addQuery() {
+        if (isLoadingRevisions.value) {
+            return;
+        }
+
+        if (props.editRouteQuery) {
+            router.push({
+                query: {
+                    ...route.query,
+                    revisionLeft: sortedRevisions.value[revisionLeftIndex.value].revision,
+                    revisionRight: sortedRevisions.value[revisionRightIndex.value].revision
+                }
+            });
+        }
+    }
+
+    function options(excludeRevisionIndex: number | undefined) {
+        return sortedRevisions.value
+            .filter((_, index) => index !== excludeRevisionIndex)
+            .map(({revision}) => ({value: revisionIndex(revision.toString()), text: revision + (currentRevisionWithSource.value.revision === revision ? ` (${t("current")})` : "")}));
+    }
+
+    const leftOptions = computed(() => {
+        return options(revisionRightIndex.value);
+    });
+
+    const rightOptions = computed(() => {
+        return options(revisionLeftIndex.value);
+    });
+
+    async function loadRevisionContent(index: number | undefined) {
+        if (index === undefined) {
+            return undefined;
+        }
+
+        const revisionObject = sortedRevisions.value[index];
+        let source = revisionObject.source;
+
+        if (!source) {
+            source = await props.revisionSource(revisionObject.revision);
+            revisionObject.source = source;
+        }
+
+        return source;
+    }
+
+    watch(revisionLeftIndex, async (newValue) => {
+        isLoadingRevisions.value = true;
+        try {
+            revisionLeftText.value = await loadRevisionContent(newValue);
+        } finally {
+            isLoadingRevisions.value = false;
+        }
+    });
+
+    watch(revisionRightIndex, async (newValue) => {
+        isLoadingRevisions.value = true;
+        try {
+            revisionRightText.value = await loadRevisionContent(newValue);
+        } finally {
+            isLoadingRevisions.value = false;
+        }
+    });
+
+    watch(() => route.query,
+          (newQuery, oldQuery) => {
+              if (newQuery.revisionLeft !== oldQuery.revisionLeft && newQuery.revisionLeft) {
+                  const newLeftIndex = revisionIndex(newQuery.revisionLeft.toString());
+                  if (newLeftIndex !== revisionLeftIndex.value) {
+                      revisionLeftIndex.value = newLeftIndex;
+                  }
+              }
+
+              if (newQuery.revisionRight !== oldQuery.revisionRight && newQuery.revisionRight) {
+                  const newRightIndex = revisionIndex(newQuery.revisionRight.toString());
+                  if (newRightIndex !== revisionRightIndex.value) {
+                      revisionRightIndex.value = newRightIndex;
+                  }
+              }
+          },
+          {deep: true}
+    );
```

Comment:
I would wish for watch to be kept to a minimum as they cannot be optimized.
If you cannot keep them within `computed` values, please make sure as much as possible that they do one thing only. It makes them easier optimize and understand.

```suggestion
    watch(() => route.query.revisionLeft,
          (revisionLeft) => {
             
                  const newLeftIndex = revisionIndex(revisionLeft.toString());
                  if (newLeftIndex !== revisionLeftIndex.value) {
                      revisionLeftIndex.value = newLeftIndex;
                  }
              })
              
watch(() => route.query.revisionRight, (revisionRight) => {
                  const newRightIndex = revisionIndex(revisionRight.toString());
                  if (newRightIndex !== revisionRightIndex.value) {
                      revisionRightIndex.value = newRightIndex;
                  }
              
    });
```

Reply:
Agree, I splitted them in 2 because there's no way around compute for that one.